File: tmp/20211125/1339. 分裂二叉树的最大乘积.ts
import { BinaryTree } from '../../6_tree/力扣加加/Tree'

// 请你删除 1 条边，使二叉树分裂成两棵子树，且它们子树和的乘积尽可能大。
// 乘积 = 某个节点下所有子节点的和 *（整个树的和 - 某个节点下所有子节点的和）
const MOD = 1e9 + 7
function maxProduct(root: BinaryTree | null): number {
  const total = sum(root)

  let res = 0

  dfs(root)
  return res % MOD

  function sum(root: BinaryTree | null): number {
    if (!root) return 0
    return root.val + sum(root.left) + sum(root.right)
  }

  function dfs(root: BinaryTree | null): number {
    if (!root) return 0
    const left = dfs(root.left)
    const right = dfs(root.right)
    res = Math.max(res, left * (total - left), right * (total - right))
    return root.val + left + right
  }
}

export {}

File: 6_tree/经典题/1339. 分裂二叉树的最大乘积.py
# 请你删除 1 条边，使二叉树分裂成两棵子树，且它们子树和的乘积尽可能大。
from collections import defaultdict
from typing import Optional


MOD = int(1e9 + 7)

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        def dfs(cur: Optional[TreeNode]) -> int:
            if not cur:
                return 0
            res = dfs(cur.left) + dfs(cur.right) + cur.val
            subtreeSum[id(cur)] = res
            return res

        subtreeSum = defaultdict(int)
        dfs(root)
        allSum = subtreeSum[id(root)]
        res = max(sub * (allSum - sub) for sub in subtreeSum.values())
        return res % MOD
